libr/anal/passmanager/passmanager.inc.c

```diff
@@ -0,0 +1,193 @@
+typedef struct Pr_(TYPE_NAME) {
+        PASS_(TYPE_NAME) *p;
+        HtPP *passResults; /*<void *object, void *result>*/
+        PM_(TYPE_NAME) *parent;
+} PR_(TYPE_NAME);
+
+/*
+ * passes: <char *Pass->name, PassRunner>
+ * running: <TYPE *object,HtPP<char *passName, NULL>>. Mainly used
+ * to detect circular dependencies
+ */
+typedef struct Pm_(TYPE_NAME) {
+	HtPP *passes;
+	HtPP *running;
+	RAnal *parent;
+} PM_(TYPE_NAME);
+
+static PR_(TYPE_NAME) *PMFN_(passrunner_new)(PASS_(TYPE_NAME) *p) {
+	PR_(TYPE_NAME) *pr = R_NEW0 (PR_(TYPE_NAME));
+	if (!pr) {
+		return NULL;
+	}
+	pr->p = p;
+
+	//FIXME create better tuned hash table
+	HtPPOptions opt = { 0 };
+	pr->passResults = ht_pp_new_opt (&opt);
+	return pr;
+}
+
+static bool PMFN_(invalidatecb) (TYPE *object, const char *passname, const PR_(TYPE_NAME) *pr) {
+	void *x = NULL;
+
+	if (!pr || !pr->p) {
+		return false;
+	}
+
+	x = pr->p->invalidate (pr->parent, pr->p, object);
+	if (x) {
+		ht_pp_insert (pr->passResults, object, x);
+	} else {
+		ht_pp_delete (pr->passResults, object);
+	}
+	return true;
+}
+
+static bool PMFN_(passcache1_free) (PASS_(TYPE_NAME) *p, const TYPE *object, void *result) {
+	if (p) {
+		p->free_result (result);
+	}
+	free (result);
+	return true;
+}
+
+static bool PMFN_(passcache2_free) (PASS_(TYPE_NAME) *p, const TYPE *object, void *result) {
+	free (result);
+	return true;
+}
+
+static bool PMFN_(passrunner_free)(void *user, const char *passname, PR_(TYPE_NAME) *pr) {
+	if (pr && pr->p && pr->p->free_result) {
+		ht_pp_foreach (pr->passResults, (HtPPForeachCallback)PMFN_(passcache1_free), pr->p);
+	} else {
+		ht_pp_foreach (pr->passResults, (HtPPForeachCallback)PMFN_(passcache2_free), pr->p);
+	}
+	if (pr && pr->p && pr->p->free_pass) {
+		pr->p->free_pass (pr->p);
+	}
+	ht_pp_free (pr->passResults);
+	free (pr);
+	return true;
+}
+
+static bool PMFN_(running_free) (void *user, TYPE *object, HtPP *objects_running_passes) {
+	ht_pp_free (objects_running_passes);
+	return true;
+}
+
+R_API PM_(TYPE_NAME) *PMFN_(new) () {
+	PM_(TYPE_NAME) *pm = R_NEW0 (PM_(TYPE_NAME));
+	if (!pm) {
+		return NULL;
+	}
+	pm->passes = ht_pp_new0 ();
+	//FIXME create better tuned hash table
+	HtPPOptions opt = { 0 };
+	pm->running = ht_pp_new_opt (&opt);
+	return pm;
+}
+
+R_API void PMFN_(set_anal)(PM_(TYPE_NAME) *pm, RAnal *anal) {
+	r_return_if_fail (pm);
+	pm->parent = anal;
+}
+
+R_API RAnal *PMFN_(get_anal)(PM_(TYPE_NAME) *pm) {
+	r_return_val_if_fail (pm, NULL);
+	return pm->parent;
+}
+
+R_API void PMFN_(free)(PM_(TYPE_NAME) *pm) {
+	ht_pp_foreach (pm->passes, (HtPPForeachCallback)PMFN_(passrunner_free), NULL);
+	ht_pp_free (pm->passes);
+	ht_pp_foreach (pm->running, (HtPPForeachCallback)PMFN_(running_free), NULL);
```

Comment:
this shouldn't be needed if you create the HtPP with the right `freefn` Option. `ht_pp_free` already calls the free function on all elements.

Reply:
fixed